src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,880 @@
+        }
+
+        mir
+    }
+
+    fn should_inline(&self, callsite: CallSite<'tcx>,
+                     callee_mir: &'a Mir<'tcx>) -> bool {
+
+        let tcx = self.tcx;
+
+        // Don't inline closures that have captures
+        // FIXME: Handle closures better
+        if callee_mir.upvar_decls.len() > 0 {
+            return false;
+        }
+
+        // Don't inline calls to trait methods
+        // FIXME: Should try to resolve it to a concrete method, and
+        // only bail if that isn't possible
+        let trait_def = tcx.trait_of_item(callsite.callee);
+        if trait_def.is_some() { return false; }
+
+        let attrs = tcx.get_attrs(callsite.callee);
+        let hint = attr::find_inline_attr(None, &attrs[..]);
+
+        let hinted = match hint {
+            // Just treat inline(always) as a hint for now,
+            // there are cases that prevent inlining that we
+            // need to check for first.
+            attr::InlineAttr::Always => true,
+            attr::InlineAttr::Never => return false,
+            attr::InlineAttr::Hint => true,
+            attr::InlineAttr::None => false,
+        };
+
+        // Only inline local functions if they would be eligible for
+        // cross-crate inlining. This ensures that any symbols they
+        // use are reachable cross-crate
+        // FIXME(#36594): This shouldn't be necessary, and is more conservative
+        // than it could be, but trans should generate the reachable set from
+        // the MIR anyway, making any check obsolete.
+        if callsite.callee.is_local() {
+            // No type substs and no inline hint means this function
+            // wouldn't be eligible for cross-crate inlining
+            if callsite.substs.types().count() == 0 && !hinted {
+                return false;
+            }
+
+        }
+
+        let mut threshold = if hinted {
+            HINT_THRESHOLD
+        } else {
+            DEFAULT_THRESHOLD
+        };
+
+        // Significantly lower the threshold for inlining cold functions
+        if attr::contains_name(&attrs[..], "cold") {
+            threshold /= 5;
+        }
+
+        // Give a bonus functions with a small number of blocks,
+        // We normally have two or three blocks for even
+        // very small functions.
+        if callee_mir.basic_blocks().len() <= 3 {
+            threshold += threshold / 4;
+        }
+
+        // FIXME: Give a bonus to functions with only a single caller
+
+        let id = tcx.map.as_local_node_id(callsite.caller).expect("Caller not local");
+        let param_env = ty::ParameterEnvironment::for_item(tcx, id);
+
+        let mut first_block = true;
+        let mut cost = 0;
+
+        // Traverse the MIR manually so we can account for the effects of
+        // inlining on the CFG.
+        let mut work_list = vec![START_BLOCK];
+        let mut visited = BitVector::new(callee_mir.basic_blocks.len());
+        while let Some(bb) = work_list.pop() {
```

Comment:
As the comment says, the traversal is done manually so it can not traverse edges that won't be taken post-inlining. If I use `mir::traversal::preorder` I can't control that as easily.